src/twisted/names/test/test_names.py

```diff
@@ -469,6 +502,66 @@ def test_zoneTransferConnectionFails(self):
             resolver.lookupZone("impossible.invalid"), error.DNSLookupError
         )
 
+    def _testHelper_zoneTransferFailure(
+        self,
+        protocolClass: Type[AccumulatingProtocol],
+        expectedFailure: Type[Exception],
+        **kwargs: Any,
+    ) -> defer.Deferred:
+        factory = RecordingServerFactory.forProtocol(protocolClass)
+        nonresponsiveTCP = reactor.listenTCP(0, factory, interface="127.0.0.1")  # type: ignore[attr-defined]
+        self.addCleanup(nonresponsiveTCP.stopListening)
+        server = nonresponsiveTCP.getHost()
+        resolver = Resolver(servers=[(server.host, server.port)])
+        d: defer.Deferred = self.assertFailure(
+            resolver.lookupZone("test-domain.com", **kwargs), expectedFailure
+        )
+
+        def _waitAnd(r, fn, *args):
+            # We need to wait a moment before calling _furtherChecks
+            # for the TCP closure to work its way through the stack
+            # and show up as accumulator.closed==True.
+            d = defer.Deferred()
+            d.addCallback(fn, *args)
+            reactor.callLater(0.01, d.callback, r)
+            return d
+
+        def _furtherChecks(
+            r: T, testCase: "ServerDNSTests", factory: RecordingServerFactory
+        ) -> T:
+            testCase.assertEqual(len(factory.instances), 1)
+            accumulator = factory.instances[0]
+            testCase.assertTrue(accumulator.made)
+            testCase.assertGreater(
+                len(accumulator.data), 0, msg="Query was received by DNS server"
+            )
+            testCase.assertTrue(accumulator.closed)
+            return r
+
+        return d.addCallback(_waitAnd, _furtherChecks, self, factory)
+
+    def test_zoneTransferConnectionDrops(self) -> defer.Deferred:
+        """
+        If the TCP connection initiated by L{Resolver.lookupZone} is lost before any
+        response is received, the deferred will errback.
+        """
+        return self._testHelper_zoneTransferFailure(
+            EasilyStartledProtocol,
+            DomainError,
+        )
+
+    test_zoneTransferConnectionDrops.todo = "Bug #3428"  # type: ignore[attr-defined]
```

Comment:
_Technically_ we're not supposed to have TODOs in trunk, but I'd allow it.

Reply:
Why? :slightly_smiling_face: It could go on the ticket or in a new branch associated with the ticket...